pgmpy/factors/FactorSet.py

```diff
@@ -0,0 +1,191 @@
+#!/usr/bin/env python3
+from pgmpy.factors import Factor
+from functools import reduce
+
+
+class FactorSet:
+    r"""
+    Base class of *Factor Sets*.
+
+    A factor set provides a compact representation of  higher dimensial factor
+    :math:`\phi_1\cdot\phi_2\cdots\phi_n`
+
+    For example the factor set corresponding to factor :math:`\phi_1\cdot\phi_2` would be the union of the factors
+    :math:`\phi_1` and :math:`\phi_2` i.e. factor set :math:`\vec\phi = \phi_1 \cup \phi_2`.
+    """
+    def __init__(self, *factors_list):
+        """
+        Initialize the factor set class.
+
+        Parameters
+        ----------
+        factors_list: Factor1, Factor2, ....
+            All the factors whose product is represented by the factor set
+
+        Examples
+        --------
+        >>> from pgmpy.factors import FactorSet
+        >>> from pgmpy.factors import Factor
+        >>> phi1 = Factor(['x1', 'x2', 'x3'], [2, 3, 2], range(12))
+        >>> phi2 = Factor(['x3', 'x4', 'x1'], [2, 2, 2], range(8))
+        >>> factor_set = FactorSet(phi1, phi2)
+        """
+        from copy import deepcopy
+        if not all(isinstance(phi, Factor) for phi in factors_list):
+            raise TypeError("Input parameters must be all factors")
+        self.factors_set = set(deepcopy(factors_list))
```

Comment:
@abinashpanda Don't you think just `self.factors` would be a more appropriate name? Like when we want to access the factors of a FactorSet we would do `fset.factors`. Right now we have to do `fset.factors_set`. What do you think?